modules/light-clients/06-solomachine/types/update_test.go

```diff
@@ -180,3 +180,386 @@ func (suite *SoloMachineTestSuite) TestCheckHeaderAndUpdateState() {
+					h.Signature = suite.GetInvalidProof()
+					clientMsg = h
+				}, false,
+			},
+			{
+				"invalid timestamp in header",
+				func() {
+					h := solomachine.CreateHeader()
+					h.Timestamp--
+					clientMsg = h
+				}, false,
+			},
+			{
+				"signature uses wrong sequence",
+				func() {
+
+					solomachine.Sequence++
+					clientMsg = solomachine.CreateHeader()
+				},
+				false,
+			},
+			{
+				"signature uses new pubkey to sign",
+				func() {
+					// store in temp before assinging to interface type
+					cs := solomachine.ClientState()
+					h := solomachine.CreateHeader()
+
+					publicKey, err := codectypes.NewAnyWithValue(solomachine.PublicKey)
+					suite.NoError(err)
+
+					data := &types.HeaderData{
+						NewPubKey:      publicKey,
+						NewDiversifier: h.NewDiversifier,
+					}
+
+					dataBz, err := suite.chainA.Codec.Marshal(data)
+					suite.Require().NoError(err)
+
+					// generate invalid signature
+					signBytes := &types.SignBytes{
+						Sequence:    cs.Sequence,
+						Timestamp:   solomachine.Time,
+						Diversifier: solomachine.Diversifier,
+						DataType:    types.CLIENT,
+						Data:        dataBz,
+					}
+
+					signBz, err := suite.chainA.Codec.Marshal(signBytes)
+					suite.Require().NoError(err)
+
+					sig := solomachine.GenerateSignature(signBz)
+					suite.Require().NoError(err)
+					h.Signature = sig
+
+					clientState = cs
+					clientMsg = h
+
+				},
+				false,
+			},
+			{
+				"signature signs over old pubkey",
+				func() {
+					// store in temp before assinging to interface type
+					cs := solomachine.ClientState()
+					oldPubKey := solomachine.PublicKey
+					h := solomachine.CreateHeader()
+
+					// generate invalid signature
+					data := append(sdk.Uint64ToBigEndian(cs.Sequence), oldPubKey.Bytes()...)
+					sig := solomachine.GenerateSignature(data)
+					h.Signature = sig
+
+					clientState = cs
+					clientMsg = h
+				},
+				false,
+			},
+			{
+				"consensus state public key is nil - header",
+				func() {
+					clientState.ConsensusState.PublicKey = nil
+					clientMsg = solomachine.CreateHeader()
+				},
+				false,
+			},
+		}
+
+		for _, tc := range testCases {
+			tc := tc
+
+			suite.Run(tc.name, func() {
+				clientState = solomachine.ClientState()
+
+				// setup test
+				tc.setup()
+
+				err := clientState.VerifyClientMessage(suite.chainA.Codec, clientMsg)
+
+				if tc.expPass {
+					suite.Require().NoError(err)
+				} else {
+					suite.Require().Error(err)
+				}
+			})
+		}
+	}
+}
+
+func (suite *SoloMachineTestSuite) TestVerifyClientMessageMisbehaviour() {
+	var (
+		clientMsg   exported.Header // TODO: Update to ClientMessage interface
+		clientState *types.ClientState
+	)
+
+	// test singlesig and multisig public keys
+	for _, solomachine := range []*ibctesting.Solomachine{suite.solomachine, suite.solomachineMulti} {
+
+		testCases := []struct {
+			name    string
+			setup   func()
+			expPass bool
+		}{
+			{
+				"successful misbehaviour",
```

Comment:
Is it possible to copy over the test case `old misbehaviour is successful (timestamp is less than current consensus state)` as well? from `misbehaviour_handle_test.go`? This test case was added as apart of an early bug fix